// react-superstore/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var shouldUpdate_1 = require("./utils/shouldUpdate");
function createStore(initialStore, reducer) {
    var store = initialStore;
    var listeners = new Set();
    function useStore(selectorFn) {
        if (selectorFn === void 0) { selectorFn = function (store) { return store; }; }
        var subscribe = (0, react_1.useCallback)(function (updater) {
            var listener = {
                updater: updater,
                selectorFn: selectorFn,
            };
            listeners.add(listener);
            return function () {
                listeners.delete(listener);
            };
        }, []);
        var syncedStore = (0, react_1.useSyncExternalStore)(subscribe, getStore, getServerStore);
        return selectorFn(syncedStore);
    }
    function setStore(action) {
        var oldStore = store;
        if (reducer) {
            store = reducer(store, action);
        }
        else {
            store = action instanceof Function ? action(store) : action;
        }
        listeners.forEach(function (_a) {
            var selectorFn = _a.selectorFn, updater = _a.updater;
            var oldState = selectorFn(oldStore);
            var newState = selectorFn(store);
            if ((0, shouldUpdate_1.default)(oldState, newState))
                updater(function () { return newState; });
        });
    }
    function getStore() {
        return store;
    }
    function getServerStore() {
        return initialStore;
    }
    return [useStore, setStore, getStore];
}
exports.default = createStore;
